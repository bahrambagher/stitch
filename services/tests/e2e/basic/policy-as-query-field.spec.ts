import { print } from 'graphql';
import { gql } from 'apollo-server-core';
import { GraphQLClient } from 'graphql-request';
import {
  PolicyArgsDefinitions,
  PolicyDefinition,
  PolicyQuery,
  PolicyType,
} from '../../../src/modules/resource-repository';
import GraphQLErrorSerializer from '../../utils/graphql-error-serializer';
import { sleep } from '../../helpers/utility';
import { RegistryMutationResponse, updatePoliciesMutation } from '../../helpers/registry-request-builder';
import { getToken } from '../../helpers/get-token';

interface TestCase {
  code: string;
  args?: PolicyArgsDefinitions;
  argsLiteral?: string;
  query?: PolicyQuery;
}

const gatewayClient = new GraphQLClient('http://localhost:8080/graphql');
const registryClient = new GraphQLClient('http://localhost:8090/graphql');

const testCases: [string, TestCase][] = [
  [
    'no_args',
    {
      code: `
        default allow = true
      `,
    },
  ],
  [
    'mandatory_arg',
    {
      code: `
        default allow = false
        allow {
          input.args.arg1
        }
      `,
      args: {
        arg1: {
          type: 'Boolean!',
        },
      },
      argsLiteral: '(arg1: true)',
    },
  ],
  [
    'mandatory_arg_with_default',
    {
      code: `
        default allow = false
        allow {
          input.args.arg1
        }
      `,
      args: {
        arg1: {
          type: 'Boolean!',
          default: '{true}',
        },
      },
    },
  ],
  [
    'optional_arg',
    {
      code: `
        default allow = false
        allow {
          input.args.arg1
        }
        allow {
          input.args.arg1 == null
        }
      `,
      args: {
        arg1: {
          type: 'Boolean',
        },
      },
      argsLiteral: '(arg1: true)',
    },
  ],
  [
    'optional_arg_not_set',
    {
      code: `
        default allow = false
        allow {
          input.args.arg1
        }
        allow {
          input.args.arg1 == null
        }
      `,
      args: {
        arg1: {
          type: 'Boolean',
        },
      },
    },
  ],
  [
    'optional_arg_with_default',
    {
      code: `
        default allow = false
        allow {
          input.args.arg1
        }
        allow {
          input.args.arg1 == null
        }
      `,
      args: {
        arg1: {
          type: 'Boolean',
          default: '{true}',
        },
      },
    },
  ],
  [
    'mandatory_args_with_default_and_without',
    {
      code: `
        default allow = false
        allow {
          input.args.arg1
          input.args.arg2
        }
      `,
      args: {
        arg1: {
          type: 'Boolean!',
          default: '{true}',
        },
        arg2: {
          type: 'Boolean!',
        },
      },
      argsLiteral: '(arg2: true)',
    },
  ],
  [
    'mandatory_arg_with_default_nested',
    {
      code: `
        default allow = false
        allow {
          input.query.policy.e2e_policies_as_query_field___mandatory_args_with_default_and_without.allow
        }
      `,
      args: {
        arg2: {
          type: 'Boolean!',
        },
      },
      query: {
        gql: print(gql`
          query($arg2: Boolean!) {
            policy {
              e2e_policies_as_query_field___mandatory_args_with_default_and_without(arg2: $arg2) {
                allow
              }
            }
          }
        `),
        variables: {
          arg2: '{arg2}',
        },
      },
      argsLiteral: '(arg2: true)',
    },
  ],
];

describe('Policies as query field', () => {
  const namespace = 'e2e_policies_as_query_field';

  beforeAll(async () => {
    expect.addSnapshotSerializer(GraphQLErrorSerializer);

    const accessToken = await getToken();
    gatewayClient.setHeader('Authorization', `Bearer ${accessToken}`);
  });

  test('Setup policies', async () => {
    const policies: PolicyDefinition[] = testCases.map(([name, { code, args, query }]) => ({
      metadata: {
        namespace,
        name,
      },
      type: PolicyType.opa,
      code,
      args,
      query,
    }));

    const policiesResponse = await registryClient.request<RegistryMutationResponse>(updatePoliciesMutation, {
      policies,
    });
    expect(policiesResponse.result.success).toBeTruthy();

    // Wait for gateway to update
    await sleep(Number(process.env.WAIT_FOR_REFRESH_ON_GATEWAY) | 1500);
  });

  test.each(testCases)('%s', async (name, { argsLiteral = '' }) => {
    const query = print(gql`
      query CheckPolicy {
        policy {
          ${namespace}___${name}${argsLiteral} {
            allow
          }
        }
      }
    `);

    const response = await gatewayClient.request(query);
    expect(response).toMatchSnapshot();
  });
});
