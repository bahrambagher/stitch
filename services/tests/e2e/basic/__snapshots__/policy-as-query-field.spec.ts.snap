// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Policies as query field mandatory_arg 1`] = `
Object {
  "policy": Object {
    "e2e_policies_as_query_field___mandatory_arg": Object {
      "allow": true,
    },
  },
}
`;

exports[`Policies as query field mandatory_arg_with_default 1`] = `
Object {
  "policy": Object {
    "e2e_policies_as_query_field___mandatory_arg_with_default": Object {
      "allow": true,
    },
  },
}
`;

exports[`Policies as query field mandatory_arg_with_default_nested 1`] = `
Object {
  "policy": Object {
    "e2e_policies_as_query_field___mandatory_arg_with_default_nested": Object {
      "allow": true,
    },
  },
}
`;

exports[`Policies as query field no_args 1`] = `
Object {
  "policy": Object {
    "e2e_policies_as_query_field___no_args": Object {
      "allow": true,
    },
  },
}
`;

exports[`Policies as query field optional_arg 1`] = `
Object {
  "policy": Object {
    "e2e_policies_as_query_field___optional_arg": Object {
      "allow": true,
    },
  },
}
`;

exports[`Policies as query field optional_arg_not_set 1`] = `
Object {
  "policy": Object {
    "e2e_policies_as_query_field___optional_arg_not_set": Object {
      "allow": true,
    },
  },
}
`;

exports[`Policies as query field optional_arg_with_default 1`] = `
Object {
  "policy": Object {
    "e2e_policies_as_query_field___optional_arg_with_default": Object {
      "allow": true,
    },
  },
}
`;
